2단계 - @MVC 구현

willing1026 opened the pull request:
`next.controller` 에 존재하는 컨트롤러들도 애노테이션 기반으로 수정했는데, 요구사항을 제대로 수행한 건지 확인 부탁드립니다.
(그리고 요구사항을 제대로 수행한건지 몰라서 불필요한 클래스들을 삭제하지 않고  `@Deprecated` 애노테이션만 추가했습니다.)

그리고 TDD방식 개발 또는 테스트케이스를 잡는 팁이 있으면 알려주실 수 있으세요? 😢 
갈수록 어떤것들을 테스트로 만들어야하는지 잘 모르겠고, TDD방식으로 개발한다는건 더 어려운거 같습니다. TDD과제들은 뭘 해야겠다는게 좀 보였는데 이번과정은 어려운거 같습니다ㅠㅠ

피드백부탁드립니다. 감사합니다!!

willing1026 committed:
# modify : "test" Method 찾는 로직 수정

- Junit3TestRunner : filter부분 getName, method명 판단 분리

willing1026 committed:
# modify : 요구사항1 기능구현

- AnnotationHandlerMapping : @Controller, @RequestMapping 어노테이션을 스캔하여 실행할 method정보를 url과 매핑하여 등록
- HandlerExecution : 기능 수행할 객체와 method정보 등록, 실행 기능 추가

willing1026 committed:
# add & modify : RequestMapping -> AnnotatedHanlderMapping 방식으로 변경중

- DispatcherServlet :
 - annotatedHandlerMapping 초기화 추가
 - RequestMapping 객체에서 controller 찾지 못할 경우 annotatedHandlerMapping에서 찾도록 기능 추가

- RequestMapping : annotatedHandlerMapping 방식으로 변경한 대상 주석처리
- *Controller : UserController로 통합, @Deprecated (삭제예정) 어노테이션 추가
- UserController : user와 관련된 여러 Controller 들을 하나의 Controller로 통합. 어노테이션 기반 method 구현
- ResultView : path에 따라 forward, redirect 방식 결정

willing1026 committed:
# add & modify : RequestMapping -> AnnotationMapping 방식 변경완료

## modify : requestMapping 관련 기능 제거
- DispatherServlet : requestMapping 인스턴스 변수 제거, controller 호출방식 annotation기반 방식으로 변경
- ForwardController, RequestMapping : @Deprecated 추가, 매핑정보 주석처리
- HomeController : 어노테이션 추가, 기능수정
- UserController : 각 method별 기능 Service 객체로 분리

## add
- UserService : UserController에 있던 기능들 서비스 객체로 분리
- ModelAndViewHandler : modelAndView 객체 생성기능 분리

javajigi commented:
@willing1026 
기존 클린코드 과정은 순수 자바이다보니 그나마 테스트하기 수월했는데요.
이번 과정은 웹, DB와 종속된 부분이 있어 쉽지 않는 것이 맞습니다.

이 단계를 TDD로 접근하려면 테스트 가능한 부분과 테스트하기 어려운 부분을 분리해 구현하도록 설계를 해야하고, 이를 위해 DI도 적극적으로 사용할 때 가능합니다. 
지금 단계에서 이를 도전하기는 쉽지 않을텐데요. 할 수 있는 만큼 테스트 가능한 부분을 분리하고, 가능한 부분에 대해서만 테스트하는 연습을 해보면 좋을 것 같아요. 100% TDD로 구현하기는 정말 힘들 겁니다.

javajigi commented on src/main/java/core/mvc/ResultView.java:
```diff
@@ -0,0 +1,27 @@
+import javax.servlet.http.HttpServletResponse;
+import java.util.Map;
+
+public class ResultView implements View {
```
향후 다양한 View 기술을 활용할 가능성이 있음.
예를 들어 View는 pdf 반환, 엑셀 반환과 같이..
그렇다면 이 객체는 ResultView보다는 JspView와 같은 이름을 사용하는 것은 어떨까?

javajigi commented on src/main/java/core/mvc/asis/DispatcherServlet.java:
```diff
@@ -15,39 +18,27 @@
     private static final long serialVersionUID = 1L;
     private static final Logger logger = LoggerFactory.getLogger(DispatcherServlet.class);
-    private static final String DEFAULT_REDIRECT_PREFIX = "redirect:";
 
```
이번 미션의 목표는 점진적으로 MVC를 전환해 보는 경험을 하는 것이 목표임
기존의 RequestMapping과 AnnotationHandlerMapping이 같이 공존해서 동작하도록 하는 것이 목표인 만큼 RequestMapping도 동작하도록 구현해 본다.

willing1026 replied:
변경하겠습니다!

javajigi replied:
이 피드백이 반영되지 않아 아쉬움이 있네요.

willing1026 replied:
앗 죄송합니다 😭 
STEP3 힌트 참고해서 다시 해보겠습니다 감사합니다!!

javajigi commented on src/main/java/core/mvc/asis/RequestMapping.java:
```diff
@@ -7,26 +7,27 @@
     private static final Logger logger = LoggerFactory.getLogger(DispatcherServlet.class);
     private Map<String, Controller> mappings = new HashMap<>();
 
     void initMapping() {
```
이곳에도 일부 Controller가 서비스 가능한 상태여야 한다.

willing1026 replied:
수정했습니다!

javajigi commented on src/main/java/next/controller/UserController.java:
```diff
@@ -0,0 +1,69 @@
+import javax.servlet.http.HttpServletResponse;
+
+@Controller
+public class UserController {
```
구현 👍

javajigi commented on src/main/java/next/service/UserService.java:
```diff
@@ -0,0 +1,93 @@
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+
+public class UserService {
```
이 클래스에 구현한 내용은 Service의 역할에 적합해 보이지 않음.
이곳의 구현은 Controller가 담당하는 것이 맞는 것으로 보임. 
모든 로직을 Controller로 이동하고 Service 클래스를 제거한다.

willing1026 replied:
@javajigi 
Controller로 기능 옮기고 Service는 제거 완료했습니다!
Controller에서 이 기능들을 담당하는게 적절한 이유를 여쭤봐도 될까요?
Service 안에서 User객체와 관련된 기능 말고도 ModelAndView 를 생성해서 반환하거나 세션정보를 해제하는 그런 기능들이 섞여있어서 적합하지 않았던건지 궁금합니다!

javajigi replied:
@willing1026
Service의 역할은 View와 의존관계를 가지지 않도록 않아야 하는데 ModelAndView를 생성하는 것은 View와 의존관계를 가지기 때문입니다.
Service의 역할에 대해 다시 한번 학습하는 계기가 되면 좋을 것 같아요.

javajigi commented on src/main/java/core/mvc/asis/DispatcherServlet.java:
```diff
@@ -15,39 +18,27 @@
 public class DispatcherServlet extends HttpServlet {
```
DispatcherServlet에 대한 테스트 코드도 구현해 본다.
다음 예제 코드와 같이 Spring Test가 제공하는 MockHttpServletRequest, MockHttpServletResponse을 활용해 테스트할 수 있다.

```java
    private DispatcherServlet dispatcher;
    private MockHttpServletRequest request;
    private MockHttpServletResponse response;

    @BeforeEach
    void setUp() {
        dispatcher = new DispatcherServlet();
        dispatcher.init();

        request = new MockHttpServletRequest();
        response = new MockHttpServletResponse();
    }
```

willing1026 replied:
DispatcherServlet 테스트 코드 추가했습니다!

javajigi reviewed (changes requested):
@MVC 구현은 잘 했는데요. 👍 
pr comment에도 남겼지만 요구사항을 잘못 이해한 것 같아요.
Legacy MVC와 @MVC가 같이 동작 가능하도록 구현해 보는 것이 목표입니다.
질문에 대한 답변 남겼으니 참고해 보세요.

willing1026 committed:
# modify & delete : 피드백 반영 ( Legacy MVC와 @MVC 같이 동작하도록 변경 )

## delete
- UserService : 제거, UserController로 기능 이관

## modify
- UserController : user와 관련된 기능 구현 ( 회원목록, 회원가입, 로그인, 로그아웃, 프로필 확인 )
- JspView : ResultView 이름 변경 ( 역할에 맞는 이름으로 변경 )
- RequestMapping, *Controller 파일들 : 주석제거, 원복

willing1026 committed:
# add : dispatcherServlet 테스트 케이스 작성

- Legacy MVC, Annotatiion MVC 테스트 케이스 작성

javajigi commented on src/main/java/core/mvc/asis/RequestMapping.java:
```diff
@@ -7,27 +7,26 @@
+        mappings.put("/users/logout", new LogoutController());
+        mappings.put("/users/create", new CreateUserController());
+        mappings.put("/users/updateForm", new UpdateFormUserController());
+        mappings.put("/users/update", new UpdateUserController());
```
이곳에 설정한 몇몇 기능들은 AnnotationHandlerMapping이로 이전했기 때문에 굳이 이곳 설정에 남아 있을 필요는 없지 않을까?

willing1026 replied:
아 정리했습니다 감사합니다 😄

javajigi commented on src/main/java/core/mvc/asis/DispatcherServlet.java:
```diff
@@ -34,11 +46,33 @@ protected void service(HttpServletRequest req, HttpServletResponse resp) throws
+            }
+
+            Controller controller = rm.findController(requestUri);
+            String viewName = controller.execute(req, resp);
```
프레임워크를 구현하는 관점에서 봤을 때 Servlet, Struts1, Struts2와 같은 다양한 프레임워크와의 통합을 하고 싶다.
이와 같이 새로운 기술에 대한 통합을 시도할 경우 if/else가 계속해서 추가될 수 있다.
이와 같이 if/else가 늘어날 것으로 예상되는 부분에 자바의 다형성을 활용해 interface를 추가하고 확장 가능한 구조로 개선해 보면 어떨까?

```
public interface HandlerAdapter {
    boolean supports(Object handler);

    ModelAndView handle(
        HttpServletRequest request, 
        HttpServletResponse response, 
        Object handler) throws Exception;
}
```

willing1026 replied:
개선했습니다!

javajigi reviewed (approved):
피드백 반영하느라 수고했네요. 👍 
일부 피드백이 반영되지 않은 내용과 추가 피드백 좀 남겼어요.
3단계 힌트와 피드백 내용 참고해 추가 리팩토링 진행해 보면 좋을 것 같아요
